### PR TITLE
Updated max_pool2d for same size tensors

### PR DESCRIPTION
**Pull request by izdeby**
Benchmark setup: 
C = 256
H/W = 64
Variance = 0

Benchmark results(CPU): 
#### Master: 
N = 8
('name', 'max_pool2d_tensor_pad'),('runs', 2867)
('name', 'max_pool2d_nt'),('runs', **362**)

N = 16
('name', 'max_pool2d_tensor_pad'),('runs', 2222)
('name', 'max_pool2d_nt'),('runs', **180**)

#### This change:
N = 8
('name', 'max_pool2d_tensor_pad'),('runs', 2869)
('name', 'max_pool2d_nt'),('runs', **2347**)

N = 16
('name', 'max_pool2d_tensor_pad'),('runs', 2214)
('name', 'max_pool2d_nt'),('runs', **1639**)

### CODE REVIEW

File: nestedtensor/csrc/functions.cpp [hunk 1]
```diff
@@ -95,8 +95,34 @@ Tensor NestedTensor_max_pool2d(
...
+
+    auto res = at::max_pool2d(at::stack(tensors),
+                              kernel_size,
+                              stride,
+                              padding,
+                              dilation,
+                              ceil_mode);
+
+    return at::detail::make_tensor<NestedTensorImpl>(
+      torch::nested_tensor::NestedTensor(std::move(res)).to_nested_tensor(nt.nested_dim() - 1));
```
Comment (izdeby) [inline nestedtensor/csrc/functions.cpp:123]: Im concerned about `nt.nested_dim() - 1`

File: nestedtensor/csrc/functions.cpp [hunk 2]
```diff
@@ -95,8 +95,34 @@ Tensor NestedTensor_max_pool2d(
     IntArrayRef dilation,
     bool ceil_mode) {
   auto self_impl = get_nested_tensor_impl(self);
-  auto self_data = self_impl->_data;
-  auto structure = self_data.get_structure();
+  auto nt = self_impl->_data;
+  auto tensor_node = nt.get_structure();
+
+  bool all_sizes_same = true;
```
Comment (cpuhrsch) [inline nestedtensor/csrc/functions.cpp:101]: I think this kind of pattern will pop up more frequently. Maybe we need a method like "is_tensor_shape"?

File: nestedtensor/csrc/functions.cpp [hunk 3]
```diff
@@ -95,8 +95,34 @@ Tensor NestedTensor_max_pool2d(
...
+    }
+  }
+ 
+  if (all_sizes_same) {
+    std::vector<at::Tensor> tensors;
+    for (auto tn : tensor_node.unbind()) {
+      tensors.push_back(tn.payload());
+    }
+
+    auto res = at::max_pool2d(at::stack(tensors),
```
Comment (cpuhrsch) [inline nestedtensor/csrc/functions.cpp:115]: You can just use the buffer and reshape it if it's contiguous

Comment (cpuhrsch) [reply nestedtensor/csrc/functions.cpp:115]: No need for stack

File: nestedtensor/csrc/functions.cpp [hunk 4]
```diff
@@ -80,6 +80,42 @@ Tensor NestedTensor_max_pool2d(
     IntArrayRef padding,
     IntArrayRef dilation,
     bool ceil_mode) {
+  auto self_impl = get_nested_tensor_impl(self);
+  auto nt = self_impl->_data;
+  auto tensor_node = get_nested_tensor_structure(self);
+
+  if (is_tensor_shape(self)) {
```
Comment (cpuhrsch) [inline nestedtensor/csrc/functions.cpp:87]: You can make this a method, but I can do that later on as part of further work

File: nestedtensor/csrc/functions.cpp [hunk 5]
```diff
@@ -80,6 +80,42 @@ Tensor NestedTensor_max_pool2d(
...
+
+      auto res = at::max_pool2d(tensor,
+                                kernel_size,
+                                stride,
+                                padding,
+                                dilation,
+                                ceil_mode);
+
+      return at::detail::make_tensor<NestedTensorImpl>(
+        torch::nested_tensor::NestedTensor(std::move(res)).to_nested_tensor(nt.nested_dim() - 1));
```
Comment (cpuhrsch) [inline nestedtensor/csrc/functions.cpp:100]: There's a helper for wrapping a C++ NestedTensor called ```wrap_nested_tensor```

**Pull request merged**